audit_log/services.py

```diff
@@ -0,0 +1,334 @@
+            AuditCommitMessage: The formatted audit log message.
+        """
+        commit_message_base = create_commit_message()
+        return AuditCommitMessage(
+            audit_event={
+                **commit_message_base["audit_event"],
+                "status": self.get_response_status(response)
+                or f"Unknown: {response.status_code}",
+                "operation": self.get_operation_name(request),
+                "actor": asdict(self.get_actor_data(request)),
+                "target": asdict(self.get_target(request)),
+            }
+        )
+
+    def commit_to_audit_log(self, request: HttpRequest, response: HttpResponse) -> None:
+        """
+        Commit the audit log to the logger and/or database.
+
+        Args:
+            request: The Django request object.
+            response: The Django response object.
+
+        Raises:
+            AuditLoggingDisabledError: If audit logging is disabled.
+        """
+        if not self.get_audit_logged_object_ids(request):
+            return
+
+        message = self.create_commit_message(request, response)
+        self.delete_audit_logged_object_ids(request)
+        self._commit_to_audit_log(message=message)
+
+    def delete_audit_logged_object_ids(self, request: HttpRequest) -> None:
+        """
+        Delete the audit logged object IDs from the request.
+
+        Args:
+            request: The Django request object.
+        """
+        if hasattr(request, audit_logging_settings.REQUEST_AUDIT_LOG_VAR):
+            delattr(request, audit_logging_settings.REQUEST_AUDIT_LOG_VAR)
+
+    def add_audit_logged_object_ids(
+        self, request: HttpRequest, instances: QuerySet
+    ) -> None:
+        """
+        Add audit logged object IDs to the request.
+
+        Args:
+            request: The Django request object.
+            instances: A single instance or a list/queryset of instances
+        """
+        request = getattr(request, "_request", request)
+        audit_logged_object_ids = set()
+
+        def add_instance(instance):
+            if not hasattr(instance, "pk") or not instance.pk:
+                return
+
+            audit_logged_object_ids.add(instance.pk)
+
+        if isinstance(instances, QuerySet) or isinstance(instances, list):
+            for instance in instances:
+                add_instance(instance)
+        else:
+            add_instance(instances)
+
+        if hasattr(request, audit_logging_settings.REQUEST_AUDIT_LOG_VAR):
+            getattr(request, audit_logging_settings.REQUEST_AUDIT_LOG_VAR).update(
+                audit_logged_object_ids
+            )
+        else:
+            setattr(
+                request,
+                audit_logging_settings.REQUEST_AUDIT_LOG_VAR,
+                audit_logged_object_ids,
+            )
+
+
+audit_log_service = AuditLogApiService()
+
+
+def create_api_commit_message_from_request(
+    request: HttpRequest, operation: Operation, object_ids: List[str]
```

Comment:
To match the documentation `operation: The operation performed, either an Operation enum value or a string.`:
```suggestion
    request: HttpRequest, operation: Operation | str, object_ids: List[str]
```

Reply:
I wanted this shortcut to support only the Operation. The pydoc should now be updated.